#         The World Bank EdStats Data Analysis

### __OBJECTIVE:__ we want to find the countries that have a strong potential of customers for our services

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


## Reading Data

In [2]:
data = pd.read_csv('EdStatsData.csv')
data.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2060  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 886930 entries, 0 to 886929
Data columns (total 70 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    886930 non-null  object 
 1   Country Code    886930 non-null  object 
 2   Indicator Name  886930 non-null  object 
 3   Indicator Code  886930 non-null  object 
 4   1970            72288 non-null   float64
 5   1971            35537 non-null   float64
 6   1972            35619 non-null   float64
 7   1973            35545 non-null   float64
 8   1974            35730 non-null   float64
 9   1975            87306 non-null   float64
 10  1976            37483 non-null   float64
 11  1977            37574 non-null   float64
 12  1978            37576 non-null   float64
 13  1979            36809 non-null   float64
 14  1980            89122 non-null   float64
 15  1981            38777 non-null   float64
 16  1982            37511 non-null   float64
 17  1983      

__Checking indicarotr__

In [56]:
data.groupby('Indicator Name').count()

Country Name  \
Indicator Name                                                     
Adjusted net enrolment rate, lower secondary, b...           242   
Adjusted net enrolment rate, lower secondary, f...           242   
Adjusted net enrolment rate, lower secondary, g...           242   
Adjusted net enrolment rate, lower secondary, m...           242   
Adjusted net enrolment rate, primary, both sexe...           242   
...                                                          ...   
Youth illiterate population, 15-24 years, male ...           242   
Youth literacy rate, population 15-24 years, bo...           242   
Youth literacy rate, population 15-24 years, fe...           242   
Youth literacy rate, population 15-24 years, ge...           242   
Youth literacy rate, population 15-24 years, ma...           242   

                                                    Country Code  \
Indicator Name                                                     
Adjusted net enrolment rate, lower secondary, b...           242   
Adjusted net enrolment rate, lower secondary, f...           242   
Adjusted net enrolment rate, lower secondary, g...           242   
Adjusted net enrolment rate, lower secondary, m...           242   
Adjusted net enrolment rate, primary, both sexe...           242   
...                                                          ...   
Youth illiterate population, 15-24 years, male ...           242   
Youth literacy rate, population 15-24 years, bo...           242   
Youth literacy rate, population 15-24 years, fe...           242   
Youth literacy rate, population 15-24 years, ge...           242   
Youth literacy rate, population 15-24 years, ma...           242   

                                                    Indicator Code  1970  \
Indicator Name                                                             
Adjusted net enrolment rate, lower secondary, b...             242     4   
Adjusted net enrolment rate, lower secondary, f...             242     2   
Adjusted net enrolment rate, lower secondary, g...             242     2   
Adjusted net enrolment rate, lower secondary, m...             242     2   
Adjusted net enrolment rate, primary, both sexe...             242    15   
...                                                            ...   ...   
Youth illiterate population, 15-24 years, male ...             242     3   
Youth literacy rate, population 15-24 years, bo...             242     3   
Youth literacy rate, population 15-24 years, fe...             242     3   
Youth literacy rate, population 15-24 years, ge...             242     3   
Youth literacy rate, population 15-24 years, ma...             242     3   

                                                    1971  1972  1973  1974  \
Indicator Name                                                               
Adjusted net enrolment rate, lower secondary, b...    29    29    25    29   
Adjusted net enrolment rate, lower secondary, f...    24    25    22    24   
Adjusted net enrolment rate, lower secondary, g...    24    24    22    24   
Adjusted net enrolment rate, lower secondary, m...    24    24    22    24   
Adjusted net enrolment rate, primary, both sexe...    15    17    18    19   
...                                                  ...   ...   ...   ...   
Youth illiterate population, 15-24 years, male ...     0     0     0     0   
Youth literacy rate, population 15-24 years, bo...     0     0     0     0   
Youth literacy rate, population 15-24 years, fe...     0     0     0     0   
Youth literacy rate, population 15-24 years, ge...     0     0     0     0   
Youth literacy rate, population 15-24 years, ma...     0     0     0     0   

                                                    1975  1976  ...  2060  \
Indicator Name                                                  ...         
Adjusted net enrolment rate, lower secondary, b...    22    27  ...     0   
Adjusted net enrolment rate, lower secondary, f

## Reading country data

In [158]:
data_country= pd.read_csv('EdStatsCountry.csv')
data_country.head()

Country Code   Short Name   Table Name                     Long Name  \
0          ABW        Aruba        Aruba                         Aruba   
1          AFG  Afghanistan  Afghanistan  Islamic State of Afghanistan   
2          AGO       Angola       Angola   People's Republic of Angola   
3          ALB      Albania      Albania           Republic of Albania   
4          AND      Andorra      Andorra       Principality of Andorra   

  2-alpha code   Currency Unit  \
0           AW   Aruban florin   
1           AF  Afghan afghani   
2           AO  Angolan kwanza   
3           AL    Albanian lek   
4           AD            Euro   

                                       Special Notes  \
0  SNA data for 2000-2011 are updated from offici...   
1  Fiscal year end: March 20; reporting period fo...   
2  April 2013 database update: Based on IMF data,...   
3                                                NaN   
4                                                NaN   

                      Region          Income Group WB-2 code  ...  \
0  Latin America & Caribbean  High income: nonOECD        AW  ...   
1                 South Asia            Low income        AF  ...   
2         Sub-Saharan Africa   Upper middle income        AO  ...   
3      Europe & Central Asia   Upper middle income        AL  ...   
4      Europe & Central Asia  High income: nonOECD        AD  ...   

            IMF data dissemination standard  \
0                                       NaN   
1  General Data Dissemination System (GDDS)   
2  General Data Dissemination System (GDDS)   
3  General Data Dissemination System (GDDS)   
4                                       NaN   

                            Latest population census  \
0                                               2010   
1                                               1979   
2                                               1970   
3                                               2011   
4  2011. Population figures compiled from adminis...   

                             Latest household survey  \
0                                                NaN   
1  Multiple Indicator Cluster Survey (MICS), 2010/11   
2               Malaria Indicator Survey (MIS), 2011   
3       Demographic and Health Survey (DHS), 2008/09   
4                                                NaN   

   Source of most recent Income and expenditure data  \
0                                                NaN   
1            Integrated household survey (IHS), 2008   
2            Integrated household survey (IHS), 2008   
3  Living Standards Measurement Study Survey (LSM...   
4                                                NaN   

  Vital registration complete Latest agricultural census  \
0                         Yes                        NaN   
1                         NaN                    2013/14   
2                         NaN                       2015   
3                         Yes                       2012   
4                         Yes                        NaN   

  Latest industrial data Latest trade data Latest water withdrawal data  \
0                    NaN            2012.0                          NaN   
1                    NaN            2012.0                         2000   
2                    NaN               NaN                         2005   
3                 2010.0            2012.0                         2006   
4                    NaN            2006.0                          NaN   

  Unnamed: 31  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 32 columns]

In [159]:
data_country.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 241 entries, 0 to 240
Data columns (total 32 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   Country Code                                       241 non-null    object 
 1   Short Name                                         241 non-null    object 
 2   Table Name                                         241 non-null    object 
 3   Long Name                                          241 non-null    object 
 4   2-alpha code                                       238 non-null    object 
 5   Currency Unit                                      215 non-null    object 
 6   Special Notes                                      145 non-null    object 
 7   Region                                             214 non-null    object 
 8   Income Group                                       214 non-null    object 
 9   WB-2 code 

__Create a DF with contry_code, name, region and income_group__

In [160]:

countries_income = pd.DataFrame(data_country, columns =['Country Code','Short Name','Region','Income Group'])

countries_income

Country Code    Short Name                      Region  \
0            ABW         Aruba   Latin America & Caribbean   
1            AFG   Afghanistan                  South Asia   
2            AGO        Angola          Sub-Saharan Africa   
3            ALB       Albania       Europe & Central Asia   
4            AND       Andorra       Europe & Central Asia   
..           ...           ...                         ...   
236          XKX        Kosovo       Europe & Central Asia   
237          YEM         Yemen  Middle East & North Africa   
238          ZAF  South Africa          Sub-Saharan Africa   
239          ZMB        Zambia          Sub-Saharan Africa   
240          ZWE      Zimbabwe          Sub-Saharan Africa   

             Income Group  
0    High income: nonOECD  
1              Low income  
2     Upper middle income  
3     Upper middle income  
4    High income: nonOECD  
..                    ...  
236   Lower middle income  
237   Lower middle income  
238   Upper middle income  
239   Lower middle income  
240            Low income  

[241 rows x 4 columns]

__Renaming Short Name to Name__

In [165]:

countries_income = pd.DataFrame({"Country_Code" : data_country["Country Code"].unique(), 
                                 "Name" : data_country["Short Name"], 
                                 "Region" : data_country["Region"], 
                                 "Income_group" : data_country["Income Group"]})
countries_income.to_csv("countries_income.csv")
countries_income

Country_Code          Name                      Region  \
0            ABW         Aruba   Latin America & Caribbean   
1            AFG   Afghanistan                  South Asia   
2            AGO        Angola          Sub-Saharan Africa   
3            ALB       Albania       Europe & Central Asia   
4            AND       Andorra       Europe & Central Asia   
..           ...           ...                         ...   
236          XKX        Kosovo       Europe & Central Asia   
237          YEM         Yemen  Middle East & North Africa   
238          ZAF  South Africa          Sub-Saharan Africa   
239          ZMB        Zambia          Sub-Saharan Africa   
240          ZWE      Zimbabwe          Sub-Saharan Africa   

             Income_group  
0    High income: nonOECD  
1              Low income  
2     Upper middle income  
3     Upper middle income  
4    High income: nonOECD  
..                    ...  
236   Lower middle income  
237   Lower middle income  
238   Upper middle income  
239   Lower middle income  
240            Low income  

[241 rows x 4 columns]

__Check indicators in Data__

In [61]:
data["Country Name"].describe()

count         886930
unique           242
top       Mauritania
freq            3665
Name: Country Name, dtype: object

We have 242 countries

In [9]:
data["Indicator Name"].describe()

count                                                886930
unique                                                 3665
top       EGRA: Oral Reading Fluency - Correct Words Rea...
freq                                                    242
Name: Indicator Name, dtype: object

3665 indicators

__Adding 2 data frames (data plus countries_income)__

In [166]:
data_p_country = pd.merge(data,countries_income, left_on='Country Code', right_on='Country_Code')
data_p_country.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2080  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2085  2090  2095  2100  Unnamed: 69  Country_Code        Name  Region  \
0   NaN   NaN   NaN   NaN          NaN           ARB  Arab World     NaN   
1   NaN   NaN   NaN   NaN          NaN           ARB  Arab World     NaN   
2   NaN   NaN   NaN   NaN          NaN           ARB  Arab World     NaN   
3   NaN   NaN   NaN   NaN          NaN           ARB  Arab World     NaN   
4   NaN   NaN   NaN   NaN          NaN           ARB  Arab World     NaN   

   Income_group  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  

[5 rows x 74 columns]

In [167]:
data_p_country.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 883265 entries, 0 to 883264
Data columns (total 74 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    883265 non-null  object 
 1   Country Code    883265 non-null  object 
 2   Indicator Name  883265 non-null  object 
 3   Indicator Code  883265 non-null  object 
 4   1970            72278 non-null   float64
 5   1971            35508 non-null   float64
 6   1972            35594 non-null   float64
 7   1973            35514 non-null   float64
 8   1974            35708 non-null   float64
 9   1975            87268 non-null   float64
 10  1976            37445 non-null   float64
 11  1977            37562 non-null   float64
 12  1978            37564 non-null   float64
 13  1979            36771 non-null   float64
 14  1980            89109 non-null   float64
 15  1981            38738 non-null   float64
 16  1982            37504 non-null   float64
 17  1983      

__Select interesting features__

Reduce the columns from data_p_country

In [170]:
data_c_reduced = data_p_country.filter(items=['Region', 'Country Name', 'Country Code', 'Income_group', 
                                                      'Indicator Name', 'Indicator Code', '2000','2005', '2010', 
                                                      '2015','2020'])
data_c_reduced.head()

Region Country Name Country Code Income_group  \
0    NaN   Arab World          ARB          NaN   
1    NaN   Arab World          ARB          NaN   
2    NaN   Arab World          ARB          NaN   
3    NaN   Arab World          ARB          NaN   
4    NaN   Arab World          ARB          NaN   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        2000       2005       2010  2015  2020  
0        NaN        NaN        NaN   NaN   NaN  
1        NaN        NaN        NaN   NaN   NaN  
2        NaN        NaN        NaN   NaN   NaN  
3        NaN        NaN        NaN   NaN   NaN  
4  77.245682  82.489487  85.211998   NaN   NaN

In [171]:
data_c_reduced.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 883265 entries, 0 to 883264
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Region          784310 non-null  object 
 1   Country Name    883265 non-null  object 
 2   Country Code    883265 non-null  object 
 3   Income_group    784310 non-null  object 
 4   Indicator Name  883265 non-null  object 
 5   Indicator Code  883265 non-null  object 
 6   2000            176625 non-null  float64
 7   2005            184036 non-null  float64
 8   2010            242382 non-null  float64
 9   2015            130966 non-null  float64
 10  2020            51436 non-null   float64
dtypes: float64(5), object(6)
memory usage: 80.9+ MB


__Remove data with no value in Income_group__

In [172]:
data_c_reduced = data_c_reduced[data_c_reduced.Income_group.notnull()]
data_c_reduced.head()

Region Country Name Country Code Income_group  \
91625  South Asia  Afghanistan          AFG   Low income   
91626  South Asia  Afghanistan          AFG   Low income   
91627  South Asia  Afghanistan          AFG   Low income   
91628  South Asia  Afghanistan          AFG   Low income   
91629  South Asia  Afghanistan          AFG   Low income   

                                          Indicator Name  Indicator Code  \
91625  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
91626  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
91627  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
91628  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
91629  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

       2000  2005  2010  2015  2020  
91625   NaN   NaN   NaN   NaN   NaN  
91626   NaN   NaN   NaN   NaN   NaN  
91627   NaN   NaN   NaN   NaN   NaN  
91628   NaN   NaN   NaN   NaN   NaN  
91629   NaN   NaN   NaN   NaN   NaN

In [173]:
data_c_reduced.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 784310 entries, 91625 to 883264
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Region          784310 non-null  object 
 1   Country Name    784310 non-null  object 
 2   Country Code    784310 non-null  object 
 3   Income_group    784310 non-null  object 
 4   Indicator Name  784310 non-null  object 
 5   Indicator Code  784310 non-null  object 
 6   2000            168090 non-null  float64
 7   2005            175458 non-null  float64
 8   2010            233465 non-null  float64
 9   2015            129126 non-null  float64
 10  2020            51128 non-null   float64
dtypes: float64(5), object(6)
memory usage: 71.8+ MB


__Select Countries with upper middle income and high income__

In [174]:
data_c_reduced.Income_group.unique()

array(['Low income', 'Upper middle income', 'High income: nonOECD',
       'Lower middle income', 'High income: OECD'], dtype=object)

In [176]:
incomes_to_keep = ['Upper middle income', 'High income: nonOECD', 'High income: OECD']
data_r_income = data_c_reduced[data_c_reduced.Income_group.isin(incomes_to_keep)]
data_r_income.head()

Region Country Name Country Code         Income_group  \
95290  Europe & Central Asia      Albania          ALB  Upper middle income   
95291  Europe & Central Asia      Albania          ALB  Upper middle income   
95292  Europe & Central Asia      Albania          ALB  Upper middle income   
95293  Europe & Central Asia      Albania          ALB  Upper middle income   
95294  Europe & Central Asia      Albania          ALB  Upper middle income   

                                          Indicator Name  Indicator Code  \
95290  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
95291  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
95292  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
95293  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
95294  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

            2000  2005       2010  2015  2020  
95290  84.558281   NaN        NaN   NaN   NaN  
95291  83.848213   NaN        NaN   NaN   NaN  
95292   0.983600   NaN        NaN   NaN   NaN  
95293  85.246353   NaN        NaN   NaN   NaN  
95294  95.889877   NaN  91.033867   NaN   NaN

In [178]:
data_r_income.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 476450 entries, 95290 to 868604
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Region          476450 non-null  object 
 1   Country Name    476450 non-null  object 
 2   Country Code    476450 non-null  object 
 3   Income_group    476450 non-null  object 
 4   Indicator Name  476450 non-null  object 
 5   Indicator Code  476450 non-null  object 
 6   2000            100496 non-null  float64
 7   2005            103939 non-null  float64
 8   2010            136908 non-null  float64
 9   2015            77816 non-null   float64
 10  2020            29876 non-null   float64
dtypes: float64(5), object(6)
memory usage: 43.6+ MB


In [179]:
data_r_income['Indicator Code'].describe()

count           476450
unique            3665
top       UIS.NAR.2.Q5
freq               130
Name: Indicator Code, dtype: object

__Select indicators for INTERNET data__

- Internet users (per 100 people)(IT.NET.USER.P2)
- 	Personal computers (per 100 people)(IT.CMP.PCMP.P2)

In [180]:
data_r_internet = data_r_income.loc[(data_r_income['Indicator Code'] == "IT.NET.USER.P2") |
                                                (data_r_income['Indicator Code'] == "IT.CMP.PCMP.P2")] 
data_r_internet.head()

Region    Country Name Country Code  \
96665        Europe & Central Asia         Albania          ALB   
97374        Europe & Central Asia         Albania          ALB   
100330  Middle East & North Africa         Algeria          DZA   
101039  Middle East & North Africa         Algeria          DZA   
103995         East Asia & Pacific  American Samoa          ASM   

               Income_group                       Indicator Name  \
96665   Upper middle income      Internet users (per 100 people)   
97374   Upper middle income  Personal computers (per 100 people)   
100330  Upper middle income      Internet users (per 100 people)   
101039  Upper middle income  Personal computers (per 100 people)   
103995  Upper middle income      Internet users (per 100 people)   

        Indicator Code      2000      2005  2010       2015  2020  
96665   IT.NET.USER.P2  0.114097  6.043891  45.0  63.252933   NaN  
97374   IT.CMP.PCMP.P2  0.756442  1.689543   NaN        NaN   NaN  
100330  IT.NET.USER.P2  0.491706  5.843942  12.5  38.200000   NaN  
101039  IT.CMP.PCMP.P2  0.630528  1.030597   NaN        NaN   NaN  
103995  IT.NET.USER.P2       NaN       NaN   NaN        NaN   NaN

In [181]:
data_r_internet.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260 entries, 96665 to 867024
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          260 non-null    object 
 1   Country Name    260 non-null    object 
 2   Country Code    260 non-null    object 
 3   Income_group    260 non-null    object 
 4   Indicator Name  260 non-null    object 
 5   Indicator Code  260 non-null    object 
 6   2000            212 non-null    float64
 7   2005            215 non-null    float64
 8   2010            120 non-null    float64
 9   2015            120 non-null    float64
 10  2020            0 non-null      float64
dtypes: float64(5), object(6)
memory usage: 24.4+ KB


In [182]:
codes_to_keep = ['IT.NET.USER.P2', 'IT.CMP.PCMP.P2']
data_r_IT = data_r_income[data_r_income['Indicator Code'].isin(codes_to_keep)]
data_r_IT.head()

Region    Country Name Country Code  \
96665        Europe & Central Asia         Albania          ALB   
97374        Europe & Central Asia         Albania          ALB   
100330  Middle East & North Africa         Algeria          DZA   
101039  Middle East & North Africa         Algeria          DZA   
103995         East Asia & Pacific  American Samoa          ASM   

               Income_group                       Indicator Name  \
96665   Upper middle income      Internet users (per 100 people)   
97374   Upper middle income  Personal computers (per 100 people)   
100330  Upper middle income      Internet users (per 100 people)   
101039  Upper middle income  Personal computers (per 100 people)   
103995  Upper middle income      Internet users (per 100 people)   

        Indicator Code      2000      2005  2010       2015  2020  
96665   IT.NET.USER.P2  0.114097  6.043891  45.0  63.252933   NaN  
97374   IT.CMP.PCMP.P2  0.756442  1.689543   NaN        NaN   NaN  
100330  IT.NET.USER.P2  0.491706  5.843942  12.5  38.200000   NaN  
101039  IT.CMP.PCMP.P2  0.630528  1.030597   NaN        NaN   NaN  
103995  IT.NET.USER.P2       NaN       NaN   NaN        NaN   NaN

In [183]:
data_r_IT.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260 entries, 96665 to 867024
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          260 non-null    object 
 1   Country Name    260 non-null    object 
 2   Country Code    260 non-null    object 
 3   Income_group    260 non-null    object 
 4   Indicator Name  260 non-null    object 
 5   Indicator Code  260 non-null    object 
 6   2000            212 non-null    float64
 7   2005            215 non-null    float64
 8   2010            120 non-null    float64
 9   2015            120 non-null    float64
 10  2020            0 non-null      float64
dtypes: float64(5), object(6)
memory usage: 24.4+ KB


In [184]:
data_r_IT.groupby('Indicator Code').count()

Region  Country Name  Country Code  Income_group  \
Indicator Code                                                     
IT.CMP.PCMP.P2     130           130           130           130   
IT.NET.USER.P2     130           130           130           130   

                Indicator Name  2000  2005  2010  2015  2020  
Indicator Code                                                
IT.CMP.PCMP.P2             130    95    96     0     0     0  
IT.NET.USER.P2             130   117   119   120   120     0

### Population

- Population, total(SP.POP.TOTL)
- Population, ages 15-24, total(SP.POP.1524.TO.UN)   
- Population, ages 15-64, total(SP.POP.1564.TO)
- 	Population growth (annual %)(SP.POP.GROW)
- School age population, upper secondary education, both sexes (number)(SP.SEC.UTOT.IN)
- School age population, secondary education, both sexes (number)(SP.SEC.TOTL.IN)
- School age population, tertiary education, both sexes (number)(SP.TER.TOTL.IN)



In [185]:
codes_to_keep = ['SP.POP.TOTL', 
                 'SP.POP.1524.TO.UN', 
                 'SP.POP.1564.TO', 
                 'SP.SEC.TOTL.IN', 
                 'SP.SEC.UTOT.IN',
                 'SP.TER.TOTL.IN']
data_r_pop = data_r_income[data_r_income['Indicator Code'].isin(codes_to_keep)]
data_r_pop.head()

Region Country Name Country Code         Income_group  \
97613  Europe & Central Asia      Albania          ALB  Upper middle income   
97616  Europe & Central Asia      Albania          ALB  Upper middle income   
97622  Europe & Central Asia      Albania          ALB  Upper middle income   
97772  Europe & Central Asia      Albania          ALB  Upper middle income   
97776  Europe & Central Asia      Albania          ALB  Upper middle income   

                                          Indicator Name     Indicator Code  \
97613  Population of the official age for secondary e...     SP.SEC.TOTL.IN   
97616  Population of the official age for tertiary ed...     SP.TER.TOTL.IN   
97622  Population of the official age for upper secon...     SP.SEC.UTOT.IN   
97772                      Population, ages 15-24, total  SP.POP.1524.TO.UN   
97776                      Population, ages 15-64, total     SP.POP.1564.TO   

            2000       2005       2010       2015  2020  
97613   511756.0   529872.0   402516.0   329011.0   NaN  
97616   257150.0   271360.0   274639.0   276247.0   NaN  
97622   240167.0   265994.0   178188.0   156541.0   NaN  
97772   538251.0   586049.0   608456.0   556269.0   NaN  
97776  1935652.0  1957782.0  1947666.0  2000257.0   NaN

In [186]:
data_r_pop.groupby('Indicator Code').count()

Region  Country Name  Country Code  Income_group  \
Indicator Code                                                        
SP.POP.1524.TO.UN     130           130           130           130   
SP.POP.1564.TO        130           130           130           130   
SP.POP.TOTL           130           130           130           130   
SP.SEC.TOTL.IN        130           130           130           130   
SP.SEC.UTOT.IN        130           130           130           130   
SP.TER.TOTL.IN        130           130           130           130   

                   Indicator Name  2000  2005  2010  2015  2020  
Indicator Code                                                   
SP.POP.1524.TO.UN             130   108   109   100   100     0  
SP.POP.1564.TO                130   111   111   111   110     0  
SP.POP.TOTL                   130   130   130   130   127     0  
SP.SEC.TOTL.IN                130   107   109   110   111     0  
SP.SEC.UTOT.IN                130   107   109   110   111     0  
SP.TER.TOTL.IN                130   109   109   107    84     0

In [187]:
data_r_pop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 780 entries, 97613 to 867473
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          780 non-null    object 
 1   Country Name    780 non-null    object 
 2   Country Code    780 non-null    object 
 3   Income_group    780 non-null    object 
 4   Indicator Name  780 non-null    object 
 5   Indicator Code  780 non-null    object 
 6   2000            672 non-null    float64
 7   2005            677 non-null    float64
 8   2010            668 non-null    float64
 9   2015            643 non-null    float64
 10  2020            0 non-null      float64
dtypes: float64(5), object(6)
memory usage: 73.1+ KB


### Education

- Barro-Lee: Percentage of population age 15+ with secondary schooling. Completed Secondary(BAR.SEC.CMPT.15UP.ZS)
- Barro-Lee: Percentage of population age 15+ with tertiary schooling. Completed Tertiary(BAR.TER.CMPT.15UP.ZS)
- UIS: Percentage of population age 25+ whose highest level of education is upper secondary, both sexes(UIS.EA.3.AG25T99)

- 	Enrolment in secondary education, both sexes (number)(SE.SEC.ENRL)
- 	Enrolment in tertiary education, all programmes, both sexes (number)(SE.TER.ENRL)


In [188]:
codes_to_keep = ['BAR.SEC.CMPT.15UP.ZS', 
                 'BAR.TER.CMPT.15UP.ZS',
                 'UIS.EA.3.AG25T99',
                 'SE.SEC.ENRL',
                 'SE.TER.ENRL',
                 'SE.XPD.TOTL.GB.ZS'
                 ]
data_r_edu = data_r_income[data_r_income['Indicator Code'].isin(codes_to_keep)]
data_r_edu.head()

Region Country Name Country Code         Income_group  \
95667  Europe & Central Asia      Albania          ALB  Upper middle income   
95669  Europe & Central Asia      Albania          ALB  Upper middle income   
96481  Europe & Central Asia      Albania          ALB  Upper middle income   
96494  Europe & Central Asia      Albania          ALB  Upper middle income   
96514  Europe & Central Asia      Albania          ALB  Upper middle income   

                                          Indicator Name  \
95667  Barro-Lee: Percentage of population age 15+ wi...   
95669  Barro-Lee: Percentage of population age 15+ wi...   
96481  Enrolment in secondary education, both sexes (...   
96494  Enrolment in tertiary education, all programme...   
96514  Expenditure on education as % of total governm...   

             Indicator Code          2000          2005       2010      2015  \
95667  BAR.SEC.CMPT.15UP.ZS      26.63000      32.92000      42.90       NaN   
95669  BAR.TER.CMPT.15UP.ZS       3.51000       2.35000       0.93       NaN   
96481           SE.SEC.ENRL  365664.00000  407403.00000  355871.00  315079.0   
96494           SE.TER.ENRL   40125.00000   63257.00000  122326.00  160527.0   
96514     SE.XPD.TOTL.GB.ZS      10.13703      11.35797        NaN       NaN   

       2020  
95667   NaN  
95669   NaN  
96481   NaN  
96494   NaN  
96514   NaN

In [189]:
data_r_edu.groupby('Indicator Code').count()

Region  Country Name  Country Code  Income_group  \
Indicator Code                                                           
BAR.SEC.CMPT.15UP.ZS     130           130           130           130   
BAR.TER.CMPT.15UP.ZS     130           130           130           130   
SE.SEC.ENRL              130           130           130           130   
SE.TER.ENRL              130           130           130           130   
SE.XPD.TOTL.GB.ZS        130           130           130           130   
UIS.EA.3.AG25T99         130           130           130           130   

                      Indicator Name  2000  2005  2010  2015  2020  
Indicator Code                                                      
BAR.SEC.CMPT.15UP.ZS             130    86    86    86     0     0  
BAR.TER.CMPT.15UP.ZS             130    86    86    86     0     0  
SE.SEC.ENRL                      130    93    97    96    76     0  
SE.TER.ENRL                      130    76    79    83    74     0  
SE.XPD.TOTL.GB.ZS                130    67    68    71    19     0  
UIS.EA.3.AG25T99                 130    14    35    61     0     0

In [190]:
data_r_edu.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 780 entries, 95667 to 868190
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          780 non-null    object 
 1   Country Name    780 non-null    object 
 2   Country Code    780 non-null    object 
 3   Income_group    780 non-null    object 
 4   Indicator Name  780 non-null    object 
 5   Indicator Code  780 non-null    object 
 6   2000            422 non-null    float64
 7   2005            451 non-null    float64
 8   2010            483 non-null    float64
 9   2015            169 non-null    float64
 10  2020            0 non-null      float64
dtypes: float64(5), object(6)
memory usage: 73.1+ KB


### Expenditures


- 	Government expenditure on education as % of GDP (%)(SE.XPD.TOTL.GD.ZS)
- Government expenditure on education not specified by level, constant US dollar (millions)(UIS.X.USCONST.UK.FSGOV)
- Capital expenditure as % of total expenditure in public institutions (%)(UIS.XSPENDP.FDPUB.FNCAP)
- Expenditure on education as % of total government expenditure (%)(SE.XPD.TOTL.GB.ZS)
- GDP per capita, PPP (current international dollar)(NY.GDP.PCAP.PP.CD)


In [191]:
codes_to_keep = ['SE.XPD.TOTL.GD.ZS', 
                 'UIS.X.USCONST.UK.FSGOV',
                 'UIS.XSPENDP.FDPUB.FNCAP',
                 'SE.XPD.TOTL.GB.ZS',
                'NY.GDP.PCAP.PP.CD']
data_r_eco = data_r_income[data_r_income['Indicator Code'].isin(codes_to_keep)]
data_r_eco.head()

Region Country Name Country Code  \
95803       Europe & Central Asia      Albania          ALB   
96514       Europe & Central Asia      Albania          ALB   
96537       Europe & Central Asia      Albania          ALB   
96550       Europe & Central Asia      Albania          ALB   
99468  Middle East & North Africa      Algeria          DZA   

              Income_group                                     Indicator Name  \
95803  Upper middle income  Capital expenditure as % of total expenditure ...   
96514  Upper middle income  Expenditure on education as % of total governm...   
96537  Upper middle income      GDP per capita, PPP (current international $)   
96550  Upper middle income  Government expenditure on education as % of GD...   
99468  Upper middle income  Capital expenditure as % of total expenditure ...   

                Indicator Code         2000         2005         2010  \
95803  UIS.XSPENDP.FDPUB.FNCAP          NaN          NaN          NaN   
96514        SE.XPD.TOTL.GB.ZS    10.137030    11.357970          NaN   
96537        NY.GDP.PCAP.PP.CD  4026.537422  5942.870083  9646.615727   
96550        SE.XPD.TOTL.GD.ZS     3.287190     3.237910          NaN   
99468  UIS.XSPENDP.FDPUB.FNCAP          NaN          NaN          NaN   

               2015  2020  
95803           NaN   NaN  
96514           NaN   NaN  
96537  11449.094589   NaN  
96550           NaN   NaN  
99468           NaN   NaN

In [192]:
data_r_eco.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 520 entries, 95803 to 866200
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          520 non-null    object 
 1   Country Name    520 non-null    object 
 2   Country Code    520 non-null    object 
 3   Income_group    520 non-null    object 
 4   Indicator Name  520 non-null    object 
 5   Indicator Code  520 non-null    object 
 6   2000            290 non-null    float64
 7   2005            295 non-null    float64
 8   2010            309 non-null    float64
 9   2015            148 non-null    float64
 10  2020            0 non-null      float64
dtypes: float64(5), object(6)
memory usage: 48.8+ KB


In [193]:
data_r_eco.groupby('Indicator Code').count()

Region  Country Name  Country Code  Income_group  \
Indicator Code                                                              
NY.GDP.PCAP.PP.CD           130           130           130           130   
SE.XPD.TOTL.GB.ZS           130           130           130           130   
SE.XPD.TOTL.GD.ZS           130           130           130           130   
UIS.XSPENDP.FDPUB.FNCAP     130           130           130           130   

                         Indicator Name  2000  2005  2010  2015  2020  
Indicator Code                                                         
NY.GDP.PCAP.PP.CD                   130   109   110   110   106     0  
SE.XPD.TOTL.GB.ZS                   130    67    68    71    19     0  
SE.XPD.TOTL.GD.ZS                   130    72    71    73    22     0  
UIS.XSPENDP.FDPUB.FNCAP             130    42    46    55     1     0

In [194]:
data_r_internet = data_r_internet.drop(['2020'], axis=1)
data_r_internet.head()

Region    Country Name Country Code  \
96665        Europe & Central Asia         Albania          ALB   
97374        Europe & Central Asia         Albania          ALB   
100330  Middle East & North Africa         Algeria          DZA   
101039  Middle East & North Africa         Algeria          DZA   
103995         East Asia & Pacific  American Samoa          ASM   

               Income_group                       Indicator Name  \
96665   Upper middle income      Internet users (per 100 people)   
97374   Upper middle income  Personal computers (per 100 people)   
100330  Upper middle income      Internet users (per 100 people)   
101039  Upper middle income  Personal computers (per 100 people)   
103995  Upper middle income      Internet users (per 100 people)   

        Indicator Code      2000      2005  2010       2015  
96665   IT.NET.USER.P2  0.114097  6.043891  45.0  63.252933  
97374   IT.CMP.PCMP.P2  0.756442  1.689543   NaN        NaN  
100330  IT.NET.USER.P2  0.491706  5.843942  12.5  38.200000  
101039  IT.CMP.PCMP.P2  0.630528  1.030597   NaN        NaN  
103995  IT.NET.USER.P2       NaN       NaN   NaN        NaN

In [195]:
data_r_internet.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260 entries, 96665 to 867024
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          260 non-null    object 
 1   Country Name    260 non-null    object 
 2   Country Code    260 non-null    object 
 3   Income_group    260 non-null    object 
 4   Indicator Name  260 non-null    object 
 5   Indicator Code  260 non-null    object 
 6   2000            212 non-null    float64
 7   2005            215 non-null    float64
 8   2010            120 non-null    float64
 9   2015            120 non-null    float64
dtypes: float64(4), object(6)
memory usage: 22.3+ KB


In [196]:
data_r_internet.describe()

2000        2005        2010        2015
count  212.000000  215.000000  120.000000  120.000000
mean    15.461754   28.933336   50.107048   66.707587
std     15.980944   24.101717   23.533872   20.328587
min      0.105046    0.564518    2.500000   12.400000
25%      3.700780    9.127204   33.225000   54.114325
50%      8.116921   21.500000   46.700000   69.709668
75%     22.943215   46.405000   69.212500   79.632177
max     75.827802   87.728023   93.390000   98.323610

In [197]:
data_r_internet['2000']= data_r_internet['2000'].fillna(data_r_internet['2000'].mode()[0])
data_r_internet['2005']= data_r_internet['2005'].fillna(data_r_internet['2005'].mean())
data_r_internet['2010']= data_r_internet['2010'].fillna(data_r_internet['2010'].mean())
data_r_internet['2015']= data_r_internet['2015'].fillna(data_r_internet['2015'].mean())


In [198]:
data_r_internet.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260 entries, 96665 to 867024
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          260 non-null    object 
 1   Country Name    260 non-null    object 
 2   Country Code    260 non-null    object 
 3   Income_group    260 non-null    object 
 4   Indicator Name  260 non-null    object 
 5   Indicator Code  260 non-null    object 
 6   2000            260 non-null    float64
 7   2005            260 non-null    float64
 8   2010            260 non-null    float64
 9   2015            260 non-null    float64
dtypes: float64(4), object(6)
memory usage: 22.3+ KB


In [200]:
data_r_internet['2000'] = data_r_internet['2000'].round(1)
data_r_internet = data_r_internet.sort_values(by='2015', ascending=False)

data_r_internet['2005'] = data_r_internet['2005'].round(1)
data_r_internet = data_r_internet.sort_values(by='2005', ascending=False)

data_r_internet['2010'] = data_r_internet['2010'].round(1)
data_r_internet = data_r_internet.sort_values(by='2010', ascending=False)

data_r_internet['2015'] = data_r_internet['2015'].round(1)
data_r_internet = data_r_internet.sort_values(by='2015', ascending=False)

#data_reduced_internet = data_reduced_internet.groupby(data_reduced_internet['Region'])
data_r_internet.head()

Region Country Name Country Code          Income_group  \
169965          North America      Bermuda          BMU  High income: nonOECD   
408190  Europe & Central Asia      Iceland          ISL     High income: OECD   
510810  Europe & Central Asia   Luxembourg          LUX     High income: OECD   
107660  Europe & Central Asia      Andorra          AND  High income: nonOECD   
624425  Europe & Central Asia       Norway          NOR     High income: OECD   

                         Indicator Name  Indicator Code  2000  2005  2010  \
169965  Internet users (per 100 people)  IT.NET.USER.P2  42.9  65.4  84.2   
408190  Internet users (per 100 people)  IT.NET.USER.P2  44.5  87.0  93.4   
510810  Internet users (per 100 people)  IT.NET.USER.P2  22.9  70.0  90.6   
107660  Internet users (per 100 people)  IT.NET.USER.P2  10.5  37.6  81.0   
624425  Internet users (per 100 people)  IT.NET.USER.P2  52.0  82.0  93.4   

        2015  
169965  98.3  
408190  98.2  
510810  97.3  
107660  96.9  
624425  96.8

In [201]:
data_r_internet.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260 entries, 169965 to 111325
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          260 non-null    object 
 1   Country Name    260 non-null    object 
 2   Country Code    260 non-null    object 
 3   Income_group    260 non-null    object 
 4   Indicator Name  260 non-null    object 
 5   Indicator Code  260 non-null    object 
 6   2000            260 non-null    float64
 7   2005            260 non-null    float64
 8   2010            260 non-null    float64
 9   2015            260 non-null    float64
dtypes: float64(4), object(6)
memory usage: 22.3+ KB


In [202]:
pivotInternet = data_r_internet.pivot_table(index=['Country Name'], 
                                                  columns=['Indicator Name'], 
                                                  values= ['2000','2005','2010','2015'])
print(pivotInternet)

                                                 2000  \
Indicator Name        Internet users (per 100 people)   
Country Name                                            
Albania                                           0.1   
Algeria                                           0.5   
American Samoa                                    0.1   
Andorra                                          10.5   
Angola                                            0.1   
...                                               ...   
United Kingdom                                   26.8   
United States                                    43.1   
Uruguay                                          10.5   
Venezuela, RB                                     3.4   
Virgin Islands (U.S.)                            13.8   

                                                           \
Indicator Name        Personal computers (per 100 people)   
Country Name                                                
Albania           